### PR TITLE
[experimental] Improve lighting curve

### PR DESCRIPTION
**Pull request by numberZero**
This version adds gaussian.

### CODE REVIEW

**Comment by paramat** [comment]
Thanks.

**Comment by Fixer-007** [comment]
We need that usual torch in dark comparison.

**Comment by paramat** [comment]
Yes i'll experiment with my 'light source against white nodes in darkness' tests.

**Comment by Fixer-007** [comment]
Applied it on newest git, everything is bright, even in caves.

**Comment by paramat** [comment]
The values will need tuning, the current values are probably just random guesses.

File: src/light.cpp [hunk 1]
```diff
@@ -50,7 +54,9 @@ void set_light_table(float gamma)
 		float x = i;
 		x /= LIGHT_MAX;
 		float brightness = a * x * x * x + b * x * x + c * x;
-		brightness = powf(brightness, 1.0 / gamma);
+		float t = boost_s * (x - boost_a);
+		float boost = boost_c * std::exp(t * t);
```
Comment (paramat) [inline src/light.cpp]: `std::exp(-t * t);` ?

**Comment by paramat** [comment]
Getting weird results, the power of e may have the wrong sign, however i have also found another lighting bug that may be complicating the situation, need to address that first.

**Comment by paramat** [comment]
Ok got it working, your maths is broken.

**Comment by paramat** [comment]
This is better, i was getting weird results caused by the 2 lines near the end, which i have commented out here.
```
// mid-range boost
	const float boost_a = g_settings->getFloat("lighting_boost_strength");
	const float boost_b = g_settings->getFloat("lighting_boost_level");
	const float boost_c = g_settings->getFloat("lighting_boost_spread");
// lighting curve coefficients
	const float a = alpha + beta - 2;
	const float b = 3 - 2 * alpha - beta;
	const float c = alpha;
// gamma correction
	gamma = rangelim(gamma, 0.5, 3.0);

	for (size_t i = 0; i < LIGHT_MAX; i++) {
		float x = i;
		x /= LIGHT_MAX;
		float brightness = a * x * x * x + b * x * x + c * x;
		float devsq = (x - boost_b) * (x - boost_b);
		float boost = boost_a * std::exp(-devsq / (2.0f * boost_c * boost_c));
		brightness = powf(brightness + boost, 1.0 / gamma);
		light_LUT[i] = rangelim((u32)(255 * brightness), 0, 255);
		//if (i > 1 && light_LUT[i] <= light_LUT[i - 1])
			//light_LUT[i] = light_LUT[i - 1] + 1;
	}
	light_LUT[LIGHT_MAX] = 255;
}
```

**Comment by paramat** [comment]
To start with here's current master light curve:

screenshot_20171116_012219

**Comment by paramat** [comment]
master

^ MT master default

screenshot_20171116_033637

^ PR with
```
lighting_boost_level = 0.4
lighting_boost_spread = 0.1
lighting_boost_strength = 0.5
```

**Comment by paramat** [comment]
screenshot_20171117_010107

^ Alpha 0.0, beta 1.0, no mid boost

**Comment by paramat** [comment]
687474703a2f2f692e696d6775722e636f6d2f5653325579384e2e706e67

^ Previous curve by sofar, the mid boost appears to be at 7

**Comment by paramat** [comment]
screenshot_20171117_011751

^ With beta = 1.0 shadows are returning

**Comment by rubenwardy** [comment]
I don't understand the continuous changing of lighting. Surely it's better to have an exact `k*x^2` over distance to stop the suspension of belief.

**Comment by paramat** [comment]
A realistic inverse square law? That's never been used in MT because it would be even darker than classic MT lighting. Which is why MC and MT have used a curve closer to linear.
https://github.com/minetest/minetest/blob/5a1975a4e38caa1385cfefac299ce17018172714/src/light.cpp
"// As in minecraft, a_n+1 = a_n * 0.8"
Then c55 added a mid boost for similar reasons sofar did and i am doing now:
"/* Middle-raised variation of a_n+1 = a_n * 0.786"

**Comment by paramat** [comment]
Essentially what i will do is find light curve parameters that recreate sofar's very popular manually created lighting curve of 0.4.16 stable. Adding a gaussian mid-boost is essential for this.

**Comment by paramat** [comment]
screenshot_20171117_020255

This is the mid boost on it's own with:
lighting_boost_level = 0.5
lighting_boost_spread = 0.15
lighting_boost_strength = 1.0
It has very little influence near light levels 0 and 15. A gaussian curve is very small at 3 standard deviations from centre (3 * 0.15 = 4.5)

**Comment by paramat** [comment]
screenshot_20171117_020834

First attempt.
lighting_alpha = 0.0
lighting_beta = 1.0
lighting_boost_level = 0.5
lighting_boost_spread = 0.15
lighting_boost_strength = 0.1

**Comment by paramat** [comment]
screenshot_20171117_022755

0.4.16 stable

**Comment by paramat** [comment]
I expect numberZero is busy so i'll open a new PR for my version.

File: src/light.cpp (hunk 1 above)
Comment (numberZero) [reply src/light.cpp]: Missed that, sorry.

**Comment by numberZero** [comment]
> so i'll open a new PR for my version.

@paramat Do that, please; I have some problems with my PC currently so can’t even test anything.

**Comment by paramat** [comment]
No smooth lighting, this shows that brightness does ondeed fall per light level, smooth lighting makes that not so obvious. Also we can sample each brightness in an image editor.

screenshot_20171117_230320

^ 0.4.16 stable

a0b1st0p1lv0p5sp0p15

^ PR with
lighting_alpha = 0.0
lighting_beta = 1.0
lighting_boost_strength = 0.1
lighting_boost_level = 0.5
lighting_boost_spread = 0.15

**Comment by paramat** [comment]
New PR #6639

**Comment by paramat** [comment]
#6639 merged.

**Pull request closed without merging**